### PR TITLE
Reschedule policy parsing

### PR DESCRIPTION
**Pull request by preetapan**
Adds parsing support for reschedule policy and adds reschedule events to alloc API structs

### CODE REVIEW

File: api/tasks.go [hunk 1]
```diff
@@ -272,6 +316,41 @@ func (g *TaskGroup) Canonicalize(job *Job) {
 		g.Update.Canonicalize()
 	}
 
+	// Merge the reschedule policy from the job
+	if jr, tr := job.Reschedule != nil, g.ReschedulePolicy != nil; jr && tr {
+		jobReschedule := job.Reschedule.Copy()
+		jobReschedule.Merge(g.ReschedulePolicy)
+		g.ReschedulePolicy = jobReschedule
+	} else if jr && !job.Reschedule.Empty() {
```
Comment (dadgar) [inline api/tasks.go]: Why restrict on the jobs being empty? What if I just want to set a no restart policy at the job and have it be inherited across all task groups?

File: api/tasks_test.go [hunk 2]
```diff
@@ -268,6 +269,105 @@ func TestTaskGroup_Canonicalize_Update(t *testing.T) {
...
+		},
+		{
+			desc: "Attempts from job, default interval",
+			jobReschedulePolicy: &ReschedulePolicy{
+				Attempts: helper.IntToPtr(1),
+			},
+			taskReschedulePolicy: nil,
+			expected: &ReschedulePolicy{
+				Attempts: helper.IntToPtr(1),
+				Interval: helper.TimeToPtr(structs.DefaultBatchJobReschedulePolicy.Interval),
```
Comment (dadgar) [inline api/tasks_test.go]: Can you add a test where only one field is set via the user in both the job and task group

File: api/tasks.go (hunk 1 above)
Comment (preetapan) [reply api/tasks.go]: When looking at how the update stanza worked, I found this commit https://github.com/hashicorp/nomad/commit/6723bade37513cd14d7ad9470c6a4a75d6252cfb and at the time convinced myself that reschedule also needed this. Having thought about it more and reading the related issue, I realize now that the problem it was solving was around reading the update stanza from the job struct via the api(which we don't have for reschedule) and then applying a job update which would override a legitimate non empty update stanza. 
In the case of reschedule, we don't add the update stanza to the job's API struct, its always inherited from the job but persisted at the task group so you are right we can remove this restriction. Will fix!

File: api/tasks_test.go (hunk 2 above)
Comment (preetapan) [reply api/tasks_test.go]: Done in lines 340-352

File: api/tasks.go [hunk 3]
```diff
@@ -78,6 +79,48 @@ func (r *RestartPolicy) Merge(rp *RestartPolicy) {
...
+func (r *ReschedulePolicy) Copy() *ReschedulePolicy {
+	if r == nil {
+		return nil
+	}
+	nrp := new(ReschedulePolicy)
+	*nrp = *r
+	return nrp
+}
+
+func (r *ReschedulePolicy) Empty() bool {
```
Comment (dadgar) [inline api/tasks.go]: Delete this

**Pull request merged**